src/control/lib/cache/cache.go

```diff
@@ -0,0 +1,196 @@
+		items: make(map[string]Item),
+	}
+	return c
+}
+
+// Set caches an item under a given key.
+func (ic *ItemCache) Set(ctx context.Context, item Item) error {
+	if ic == nil {
+		return errors.New("ItemCache is nil")
+	}
+
+	if common.InterfaceIsNil(item) || item.Key() == "" {
+		return errors.New("invalid item")
+	}
+
+	ic.mutex.Lock()
+	defer ic.mutex.Unlock()
+
+	ic.set(ctx, item)
+	return nil
+}
+
+func (ic *ItemCache) set(ctx context.Context, item Item) {
+	ic.items[item.Key()] = item
+}
+
+// Delete fully deletes an Item from the cache.
+func (ic *ItemCache) Delete(key string) {
+	if ic == nil {
+		return
+	}
+
+	ic.mutex.Lock()
+	defer ic.mutex.Unlock()
+
+	delete(ic.items, key)
+}
+
+// Has checks whether any item is cached under the given key.
+func (ic *ItemCache) Has(key string) bool {
+	if ic == nil {
+		return false
+	}
+
+	ic.mutex.RLock()
+	defer ic.mutex.RUnlock()
+
+	_, found := ic.items[key]
+	return found
+}
+
+type ItemCreateFunc func() (Item, error)
+
+type errKeyNotFound struct {
+	key string
+}
+
+func (e *errKeyNotFound) Error() string {
+	return fmt.Sprintf("key %q not found", e.key)
+}
+
+func noopRelease() {}
+
+// GetOrCreate returns an item from the cache if it exists, otherwise it creates
+// the item using the given function and caches it. The item must be released
+// by the caller when it is safe to be modified.
+func (ic *ItemCache) GetOrCreate(ctx context.Context, key string, missFn ItemCreateFunc) (Item, func(), error) {
+	if ic == nil {
+		return nil, noopRelease, errors.New("nil ItemCache")
+	}
+
+	if key == "" {
+		return nil, noopRelease, errors.Errorf("empty string is an invalid key")
+	}
+
+	if missFn == nil {
+		return nil, noopRelease, errors.Errorf("item create function is required")
+	}
+
+	ic.mutex.Lock()
+	defer ic.mutex.Unlock()
+
+	item, err := ic.get(key)
+	if err != nil {
+		ic.log.Debugf("failed to get item for key %q: %s", key, err.Error())
+		item, err = missFn()
+		if err != nil {
+			return nil, noopRelease, errors.Wrapf(err, "create item for %q", key)
+		}
+		ic.log.Debugf("created item for key %q", key)
+		ic.set(ctx, item)
+	}
+
+	if item.NeedsRefresh() {
+		if err := item.Refresh(ctx); err != nil {
+			return nil, noopRelease, errors.Wrapf(err, "fetch data for %q", key)
+		}
+		ic.log.Debugf("refreshed item %q", key)
+	}
+	item.Lock()
+
+	return item, item.Unlock, nil
+}
+
+// Get returns an item from the cache if it exists, otherwise it returns an
+// error. The item must be released by the caller when it is safe to be modified.
+func (ic *ItemCache) Get(ctx context.Context, key string) (Item, func(), error) {
+	if ic == nil {
+		return nil, noopRelease, errors.New("nil ItemCache")
+	}
+
+	if key == "" {
+		return nil, noopRelease, errors.Errorf("empty string is an invalid key")
+	}
+
+	ic.mutex.RLock()
+	defer ic.mutex.RUnlock()
+
+	item, err := ic.get(key)
+	if err != nil {
+		return nil, noopRelease, err
+	}
+
+	if item.NeedsRefresh() {
+		if err := item.Refresh(ctx); err != nil {
```

Comment:
My only hesitation with this pattern (and the reason I removed `NeedsRefresh()` from the interface) is that it's potentially racy. The `NeedsRefresh()` check and `Refresh()` actions are done under different locks. It's not unsafe, as far as I can tell, but could result in a refresh being done unnecessarily if more than one goroutine does the `NeedsRefresh()` check nearly simultaneously.

Reply:
Good point, I'll revisit this.